### ISSUE TITLE
Example wanted: Logging in to legacy API without OAuth 2.0 support and parsing token to oauth_token

### ISSUE THREAD

**Issue by exetico** [r-lib/httr2#491]
Hi,

I've tried and found a way to login to a API where OAuth 2.0 are not used.
In the API, I login with username and password in the body, and I get a "token" and "expires" in return from my POST request.

I've found the [https://httr2.r-lib.org/reference/oauth_token.html?q=oauth_token#null](oauth_token) and the oauth_token_cached. But I'm not really sure how I should cook the functions together, so I can set and reuse the token, and request a new one, once expired.

After looking through the issues and the documentation, I'm still not sure how to continue from here.

Are there a good example on how I can do this?

**Comment by hadley** [comment]
Can you point to the API docs you're binding?

**Comment by exetico** [comment]
Sure. I can also see that my link are broken.

I've noticed that there's a [Authenticate request with bearer token (`req_auth_bearer_token`)](https://httr2.r-lib.org/reference/req_auth_bearer_token.html?q=request%20token#arguments), but I'd like to make a oauth like [request (basic POST)](https://httr2.r-lib.org/articles/wrapping-apis.html?q=request#basic-request), where I'm able to parse the token into [oauth_token_cached](https://httr2.r-lib.org/reference/oauth_token_cached.html?q=token_cache#null) so I'm able to use it in my httr2 requests, and renew the token, once it's expires.

A use case:
1/ `GET` to `/test` to get a API status (in my R-package)
2/ `POST` to `/login`with `{"username":"<username>","password":"<password>"}` (Normal JSON type) to fetch a token. The response does contain a JSON object with `token` and `expires`, so I do need to calculate `expires in` and parse in the token to [`oauth_token`](https://httr2.r-lib.org/reference/oauth_token.html).
3/ By now, I expects the solution to have the token set, including expire time
4/ Use the Bearer token in other requests to user-related endpoints like `/users/me`
5/ Secure renewal of the token, once it's expired

I've found the [`req_oauth`](https://httr2.r-lib.org/reference/req_oauth.html?q=oauth_token#value) which does sounds like a functions I'd use to get the token, and cache it with httr2 utils, but I'm not really sure how to get this working as expected.

I'm not really sure if that's what you want, but I hope it does make some kind of sense. I'll be using httr2 in a R-package to wrap around a API, so in general, the documentation are **brilliant'* already. It's just this use case where I need to fetch a token, [store it in cache](https://httr2.r-lib.org/articles/oauth.html?q=cache#caching), and use it, including renewal once needed.

I've already done the 1/ + 2/ to fetch the token with just the regular requests utils, but I've not managed to assign that as a httr2 token resource, to be used in 3/+4/+5/.

**Comment by hadley** [comment]
I meant docs for the web API you were trying to wrap so I could better understand what it wants. But this more detailed explanation was helpful too, thanks 😄 

Yeah, it does sound like `req_oauth()` is what you want, but unfortunately it's well documented, and I don't remember how it works without spending a few hours getting my brain back into httr2. You might be able to figure it yourself by looking at some of the other `req_oauth_` functions; otherwise I'll try and give you more insight the next time I'm working deeply on httr2.

**Comment by exetico** [comment]
I totally understand 😄! 

I'd like to mention that this is a R-lib for a API **I** manage; so I've implemented a minimum level of Oauth 2.0 now, as it's clear that my R skills are not that deep (😆), compared to other languages, at least as this moment.

With that said; Maybe other people are looking for a way, to handle this case, with a API they're not able to just change in a hour.

Before updating the API i did manage to mock a "test_con_client" fnc. with `oauth_client` and a "req_auth" fnc. with `req_oauth_password`. I also sent the "login/auth" request to our API, but I was facing three problems:

1. How could I force the req_oauth_password solution to send data as a JSON obj., and not `urlencoded`
2. How could I force the "flow" to understand that the access_token was within a `token` key (not `access_token`)
3. How could I handle a `expires_at` key-value, and convert that to a `expires_in` value to be used in httr2

I know this is a edge-case, but if it's already possible to do a thing like this in the library, I guess it would be a good help for a beginner, with something in the docs, to play around with a API with a non-standard approch to authentication.

By handling 1+2+3, or, illustrating how it's possible to parse in the token with code, like "assign this access token y to the httr2 lib, which expires_in x", or similar.

As I've implemented the required specs in the API instead, you're more than welcome to close the issue, if you find it too niche or not that important.

**Comment by exetico** [comment]
_A follow-up, now with a API which supports username+password login through OAuth 2.0 standard:_

I noticed that #119 points out that `req_oauth_password` prompts the "Password" input prompt on **all** request. Have I implemented something in a wrong way for my first "initial httr2-based project"?

I use the following simple test-code
```r
get_users_me <- function() {
  response <- NULL

  if (interactive()) {
    tryCatch({
      response <- request("https://domain.com/users/me") |>
       req_auth() |>
       req_perform()
     
     # Handle the response inside the conditional
     content <- response |> resp_body_json(auto_unbox = TRUE)
    }, error = function(e) {
      stop("An error occurred: ", conditionMessage(e))
    })
  } else {
    # Handle non-interactive case, e.g., tests or scripts
    # Make sure to define how response should be handled or mocked here
  }

  # return content
  content
}
```

And the following functions defined for the url, client and req auth:

```r
test_con_client <- function() {
  oauth_client(
    id = "28acfec0674bb3da9f39",
    token_url = "https://domain.com/oauth/token",
    name = "client-r-package"
  )
}

req_auth <- function(req) {
  req_oauth_password(req,
    client = test_con_client(),
    username = "user@domain.com"
  )
}
```

If I enable "cache_disk", it's the same thing. It prompts me every time. But it do look like it's actually using the existing token, even through it re-prompts me, asking me to enter the password again in step 1.

By running `with_verbosity(get_users_me())`, it's clear that:
1) No. 1 run does a `POST /oauth/token HTTP/2`, get's a `HTTP/2 200` back, **and** requests the resources at `GET /users/me HTTP/2` and gets another `HTTP/2 200`. The Bearer are set in the headers, as expected
2) No. 2 run (**same** R session) does only make a `GET /users/me HTTP/2` and gets a `HTTP/2 200` back.

However, the password input was prompted on both runs:

<img width="738" alt="billede" src="https://github.com/r-lib/httr2/assets/3549445/8592df05-84f0-4acc-842e-ea8d7d71c985">
_Ignore the code in the background._

In other words, it's clear to me, that the password are not used in run no. 2.

Could this be theated as a bug in the implementation, or do **I** need to check something, and only wrap `req_oauth_password`, in a case where it's needed?

If I run the official [example with github_client](https://httr2.r-lib.org/articles/oauth.html#authenticating-a-request) with [`req_oauth_auth_code`](https://httr2.r-lib.org/reference/req_oauth_auth_code.html), thinks works as I expect it to do: 1. run prompts me to grant access, and fetch the information. 2. run just fetches the information. I'm not "reprompted" to grant access.
